Property search blog (#3045)
pulumi/pulumi-hugo: themes/default/content/blog/property-search/index.md

# Lowercase, hyphen-delimited is recommended.
tags:
    - pulumi-cloud
    - features

# See the blogging docs at https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md.
# for additional details,
# and please remove these comments before submitting for review.
---

When we launched Resource Search for Pulumi Cloud in April we extended the capability of Pulumi Cloud beyond just managing infrastructure as code to managing your cloud footprint. Today, I'm excited to announce that we are taking this a step further and exposing a much larger set of data to search over: resource properties. Pulumi Property Search enables you to search on a wide array of resource properties, everything from instance type, to database version, to region and availability zone to if resources are open to the internet.
Property Search is applicable across all cloud resources, whether they reside in AWS, Google Cloud, Azure, or any other provider. Pulumi Cloud Resource Search is available on the Enterprise and Business Critical editions of Pulumi Cloud.

<!--more-->

### See it in action!

![Searching on resource properties in the Pulumi Cloud console](properties.gif)

## Feature Overview

> Cloud resources, regardless of the provider, come with various properties that define their behavior, performance, location, and more. With Pulumi Cloud Property Search, you can now search resources based on these properties. Whether you're trying to find resources running in a specific region, instances with a particular tag, or databases running a certain engine version, just like Resource Search, searching over properties is designed to streamline and simplify your tasks.

Reviewer: ```suggestion
Cloud resources, regardless of the provider, come with various properties that define their behavior, performance, location, and more. With Pulumi Cloud Property Search, you can now search resources based on these properties. Whether you're trying to find resources running in a specific region, instances with a particular tag, or databases running a certain engine version, searching over properties with Pulumi Cloud streamlines and simplifies your tasks.
```

> Using the new feature is simple and straightforward. Go to the Pulumi console, navigate to the search bar, and start typing the property you're interested in. For example, if you're searching for all AWS RDS instances running MySQL with a particular engine version, simply type .engineVersion:5.7 and press enter.

Reviewer: a screenshot to accompany this section might be helpful

Since launching Resource Search, a majority of Pulumi customers have used it, making it the most quickly adopted Pulumi feature. One of these customers is Alkria, a company that reinvents networking for the cloud era with global unified network infrastructure delivered as-a-service. “I’m making developers at Alkira significantly more productive while also making my job easier using Pulumi’s IaC platform and features like Pulumi Insights and Deployments,” said Santosh Dornal, head of software test & DevOps, Alkira. “I can get developers using IaC immediately with Pulumi Deployments and its GitHub integration, while Pulumi Insights makes it really easy to find idle developer environments that need to be shut down, which reduces our cloud costs.”

## Example queries

### AI Assist & properties

> Our AI Assist functionality allows you to use natural language to search for resources. AI Assist works by converting natural language to a query in our search syntax. This functionality becomes even more valuable with property search. Resources can have dozens of properties, remembering exactly what the property values are can be challenging. AI Assist is going to step in so that you don’t have to.

Reviewer: ```suggestion
Our AI Assist functionality allows you to use natural language to search for resources. AI Assist works by converting natural language to a query in our search syntax. This functionality becomes even more valuable with property search. Resources can have dozens of properties, and remembering exactly how to search for property values can be challenging. AI Assist is going to step in so that you don’t have to.
```

Here is a list of example queries we have seen customers use:

- resources tagged with stack production
- my aurora engines running MySQL engine version 5
- s3 buckets read write in production
- buckets objects in all clouds
- show me virtual machines in all clouds
- all resources in us east 1

![Placeholder Image](meta.png)

### Search syntax

- `project:pulumi-service stack:staging .instanceType:`
- `type:aws:rds/cluster:Cluster .engine:aurora .engineVersion:8`
- `type:aws:s3/bucket:Bucket stack:production .acl:public-read`
- `team:data package:snowflake .availabilityZone:us-west-2a`

### Use the REST API

> Now you have all this data at your fingertips, but searching isn’t the only thing you can do with it. You can also export it to perform analytics: whether that is something simple like opening it in a spreadsheet or if you want to regularly ingest it to your data warehouse. You can download a CSV from the console or get the data using the Data Export API.

Reviewer: ```suggestion
Now that you have all this data at your fingertips, you can do much more with it. You can also export it to perform analytics: whether that is something simple like opening it in a spreadsheet or if you want to regularly ingest it to your data warehouse. You can download a CSV from the console or get the data using the Data Export API.
```

---

> title: "Property Search: Enhanced Search Capabilities on Pulumi Cloud"

Reviewer 2: Perhaps?

```suggestion
title: "Property Search: Enhanced Resource Search in Pulumi Cloud"
```

# The date represents the post's publish date,
# and by default corresponds with the date this file was generated.
# Posts with future dates are visible in development,
# but excluded from production builds.
# Use the time and timezone-offset portions of of this value
# to schedule posts for publishing later.
date: 2023-06-13T09:00:08-07:00

# Use the meta_desc property to provide a brief summary
# (one or two sentences) of the content of the post,
# which is useful for targeting search results or social-media previews.
# This field is required or the build will fail the linter test.
# Max length is 160 characters.
meta_desc: Pulumi Cloud launches Property Search, a feature on top of Resource Search that opens search to resource properties.

# The meta_image appears in social-media previews and on the blog home page.
# A placeholder image representing the recommended format, dimensions and aspect ratio
# has been provided for you.
meta_image: meta.png

# At least one author is required.
# The values in this list correspond with the `id` properties
# of the team member files at /data/team/team.
# Create a file for yourself if you don't already have one.
authors:

>     - meagan-cojocar
>     - bryce-lampe

Reviewer 3: You wrote it all! No need to credit me :)
```suggestion
    - meagan-cojocar
```

# At least one tag is required.
# Lowercase, hyphen-delimited is recommended.
tags:
    - pulumi-cloud
    - features

# See the blogging docs at https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md.
# for additional details,
# and please remove these comments before submitting for review.
---

> When we launched [Resource Search for Pulumi Cloud](/blog/pulumi-insights) in April we extended the capability of Pulumi Cloud beyond just managing infrastructure as code to managing your cloud footprint. Today, we're excited to announce that we are taking this a step further and exposing a much larger set of data to search over: resource properties. Pulumi Property Search enables you to search on a wide array of resource properties, everything from instance type, to database version, to region and availability zone, to if resources are open to the internet.

Reviewer 3: Nit. "to if" reads strangely to me, but feel free to reject.

```suggestion
When we launched [Resource Search for Pulumi Cloud](/blog/pulumi-insights) in April we extended the capability of Pulumi Cloud beyond just managing infrastructure as code to managing your cloud footprint. Today, we're excited to announce that we are taking this a step further and exposing a much larger set of data to search over: resource properties. Pulumi Property Search enables you to search on a wide array of resource properties, everything from instance type, to database version, to region and availability zone, to whether resources are open to the internet.
```

> When we launched [Resource Search for Pulumi Cloud](/blog/pulumi-insights) in April we extended the capability of Pulumi Cloud beyond just managing infrastructure as code to managing your cloud footprint. Today, we're excited to announce that we are taking this a step further and exposing a much larger set of data to search over: resource properties. Pulumi Property Search enables you to search on a wide array of resource properties, everything from instance type, to database version, to region and availability zone, to whether resources are open to the internet.

Reviewer 2: ```suggestion
Two months ago we launched [Resource Search for Pulumi Cloud](/blog/pulumi-insights) to offer multi-cloud resource search and analytics across every cloud environment in your organization. Today, we're excited to announce that we are taking this a step further and exposing a much larger set of data to search over: resource properties. Including property names and values in your resource searches enables you to search on an even richer variety of dimensions, everything from instance type, to database version, to region and availability zone, to whether resources are open to the internet.  By searching on property data, in addition to overall resource metadata, you can craft expressive queries that serve key compliance, operational and diagnostic use cases.
```

<!--more-->

> Property Search is applicable across all cloud resources, whether they reside in AWS, Google Cloud, Azure, or any of Pulumi's 100+ cloud providers. Pulumi Cloud Resource Search is available on the Enterprise and Business Critical editions of Pulumi Cloud.

Reviewer 3: > Pulumi Cloud Resource Search is available on the Enterprise and Business Critical editions of Pulumi Cloud.

Should this be "Pulumi Cloud Property Search is available..."? "Plain" Resource Search is available more broadly.

> Property Search is applicable across all cloud resources, whether they reside in AWS, Google Cloud, Azure, or any of Pulumi's 100+ cloud providers. Pulumi Cloud Resource Search is available on the Enterprise and Business Critical editions of Pulumi Cloud.

Reviewer 2: Was the second sentence not correct?  Or perhaps I was reading it wrong?

```suggestion
Pulumi Resource Search, including the ability to search on property data, is applicable across all cloud resources, whether they reside in AWS, Google Cloud, Azure, or any of Pulumi's 100+ cloud providers. Support for querying on property data is available in the Enterprise and Business Critical editions of Pulumi Cloud.
```

Also, should we link off to pricing page for edition information?

Author: > Was the second sentence not correct? Or perhaps I was reading it wrong?

It was incorrect! Fixed in my latest

### See it in action

> 

Reviewer 2: Worth having a short summary of what this gif shows?

Author: We didn't figure it out. Turns out we both haven't documented and don't give access to the bucket behind Zach Sean and Christian. Christian did offer to upload it for me but given the time constraints my latest has a short compressed gif.

> ![Searching on resource properties in the Pulumi Cloud console](properties.gif)

Reviewer 2: I don't see the gif here - did we figure out how to host it in asset storage?

Author: We didn't figure it out. Turns out we both haven't documented and don't give access to the bucket behind Zach Sean and Christian. Christian did offer to upload it for me but given the time constraints my latest has a short compressed gif.

## Feature Overview

> Cloud resources, regardless of the provider, come with various properties that define their behavior, performance, location, and more. With Pulumi Cloud Property Search, you can now search resources based on these properties. Whether you're trying to find resources running in a specific region, instances with a particular tag, or databases running a certain engine version, searching over properties with Pulumi Cloud streamlines and simplifies your tasks.

Reviewer 2: ```suggestion
Cloud resources, regardless of the provider, come with various properties that define their behavior, performance, location, and more. You can now use Pulumi Resource Search to search resources based on these properties in addition to other resource metadata. Whether you're trying to find resources running in a specific region, instances with a particular tag, or databases running a certain engine version, searching over properties with Pulumi Cloud streamlines and simplifies your tasks.
```

> Using the new feature is simple and straightforward. Go to the Pulumi console, navigate to the search bar, and start typing the property you're interested in. For example, if you're searching for all AWS RDS instances running MySQL with a particular engine version, simply type .engineVersion:5.7 and press enter.

Reviewer 2: ```suggestion
Using the new feature is simple and straightforward. Go to the Pulumi console, navigate to the search bar, and start typing the property you're interested in. For example, if you're searching for all AWS RDS instances running MySQL with a particular engine version, simply type `.engineVersion:5.7` and press enter.
```

The ability to search for resources based on their properties is a powerful tool for optimizing cloud resource management. It provides deeper insights into your infrastructure and aids in identifying resources that are misconfigured, out-of-date, or not adhering to your organization's best practices. Essentially, it's a significant step towards more proactive, informed, and efficient management of your cloud environment.

> Since launching Resource Search, a majority of Pulumi customers have used it, making it the most quickly adopted Pulumi feature. One of these customers is Alkria, a company that reinvents networking for the cloud era with global unified network infrastructure delivered as-a-service. “I’m making developers at Alkira significantly more productive while also making my job easier using Pulumi’s IaC platform and features like Pulumi Insights and Deployments,” said Santosh Dornal, head of software test & DevOps, Alkira. “I can get developers using IaC immediately with Pulumi Deployments and its GitHub integration, while Pulumi Insights makes it really easy to find idle developer environments that need to be shut down, which reduces our cloud costs.”

Reviewer 3: Nit: capitalize "head of software test & DevOps" since it's a job title?

> Since launching Resource Search, a majority of Pulumi customers have used it, making it the most quickly adopted Pulumi feature. One of these customers is Alkria, a company that reinvents networking for the cloud era with global unified network infrastructure delivered as-a-service. “I’m making developers at Alkira significantly more productive while also making my job easier using Pulumi’s IaC platform and features like Pulumi Insights and Deployments,” said Santosh Dornal, head of software test & DevOps, Alkira. “I can get developers using IaC immediately with Pulumi Deployments and its GitHub integration, while Pulumi Insights makes it really easy to find idle developer environments that need to be shut down, which reduces our cloud costs.”

Reviewer 2: ```suggestion
Since launching Resource Search, a majority of Pulumi Cloud customers have used it, making it the most quickly adopted feature of Pulumi Cloud. One of these customers is Alkria, a company that reinvents networking for the cloud era with global unified network infrastructure delivered as-a-service. “I’m making developers at Alkira significantly more productive while also making my job easier using Pulumi’s IaC platform and features like Pulumi Insights and Deployments,” said Santosh Dornal, head of software test & DevOps, Alkira. “I can get developers using IaC immediately with Pulumi Deployments and its GitHub integration, while Pulumi Insights makes it really easy to find idle developer environments that need to be shut down, which reduces our cloud costs.”
```

Do we want to move the quote into a quote block to highlight it a bit more explicitly?  Or else perhaps move it into the conclusion if it interrupts the flow?

## Example queries

> Before Property Search using just Resource Search you were able to find resources by their name, type, URN, parent, the team they are in and so on, but today any property on your resource will be indexed by Pulumi Cloud and searchable after each update. Let's walk through some examples of the types of queries you can now perform.

Reviewer 3: ```suggestion
Before Property Search, using just Resource Search, you were able to find resources by their name, type, URN, parent, the team they are in, and so on... but today, any property on your resource will be indexed by Pulumi Cloud and searchable after each stack update. Let's walk through some examples of the types of queries you can now perform.
```

> Before Property Search, using just Resource Search, you were able to find resources by their name, type, URN, parent, the team they are in, and so on... but today, any property on your resource will be indexed by Pulumi Cloud and searchable after each stack update. Let's walk through some examples of the types of queries you can now perform.

Reviewer 2: ```suggestion
Using Resource Search prior to support for property data, you were able to find resources by their name, type, URN, parent, the team they are in, and so on... but today, any property on your resource will be indexed by Pulumi Cloud and searchable after each stack update. Let's walk through some examples of the types of queries you can now perform.
```

### AI Assist & properties

> Our [AI Assist](/product/private-previews) functionality allows you to use natural language to search for resources. AI Assist works by converting natural language to a query in our search syntax. This functionality becomes even more valuable with property search. Resources can have dozens of properties, and remembering exactly how to search for property values are can be challenging, especially when they change depending on the provider. AI Assist is going to step in so that you don’t have to remember everything.

Reviewer 2: Is that the best link for AI Assist?
```suggestion
Our [AI Assist](/product/private-previews) functionality allows you to use natural language to search for resources. AI Assist works by converting natural language to a query in our search syntax. This functionality becomes even more valuable with property search. Resources can have dozens of properties, and remembering exactly how to search for property values are can be challenging, especially when they change depending on the provider. AI Assist is going to step in so that you don’t have to remember everything.
```

Here is a list of example queries we have seen customers use:

- resources tagged with stack production
- my aurora engines running MySQL engine version 5
- s3 buckets read write in production
- buckets objects in all clouds
- show me virtual machines in all clouds
- all resources in us east 1

### Search syntax

Properties can be used in your queries by using a `.` prefix. For example, if you want to see all resources with an Instance Type property, you type in the search bar `.instanceType:`. If you want to only look at a specific value, you can search for the value behind the colon. In this case our query would become `.instanceType:m5.xlarge`.

![Picture of .instanceType:m5.xlarge query in the dashboard](instance_type.png)

> Here are some query examples we have seen customers use:

Reviewer 2: Agreed on this - having english descriptions of these would be great!

- `project:pulumi-service stack:staging .instanceType:`
- `type:aws:rds/cluster:Cluster .engine:aurora .engineVersion:8`
- `type:aws:s3/bucket:Bucket stack:production .acl:public-read`
- `team:data package:snowflake .availabilityZone:us-west-2a`
- `vpc created:2023-06-12 .tags.Environment:production`

### Use the REST API

Now you have all this data at your fingertips, but searching isn’t the only thing you can do with it. You can also export it to perform analytics: whether that is something simple like opening it in a spreadsheet or if you want to regularly ingest it to your data warehouse. You can download a CSV from the console or get the data using the [Data Export API](/docs/pulumi-cloud/cloud-rest-api/#data-export).

> But the value of Resource Search and Property Search are not limited to what you do in the console and for analytics. You can also leverage the [Resource Search API](/docs/pulumi-cloud/cloud-rest-api/#resource-search) to create automation and workflows off of search results.

Reviewer 2: ```suggestion
But the value of Resource Search is not limited to what you do in the console and for analytics. You can also leverage the [Resource Search API](/docs/pulumi-cloud/cloud-rest-api/#resource-search) to create automation and workflows off of search results.
```

{{< chooser language "javascript,typescript,python,go" >}}

{{% choosable language javascript %}}

```javascript
const headers = {
  Accept: "application/json",
  Authorization: "token pul-abc123",
};

const body = await (
  await fetch("https://api.pulumi.com/api/orgs/{org}/search/resources", {
    method: "GET",
    headers: headers,
  })
).json();

console.log(body);
```

{{< /choosable >}}

{{% choosable language typescript %}}

```typescript
const headers = {
  Accept: "application/json",
  Authorization: "token pul-abc123",
};

const body = await(
  await fetch("https://api.pulumi.com/api/orgs/{org}/search/resources", {
    method: "GET",
    headers: headers,
  }),
).json();

console.log(body);
```

{{< /choosable >}}

{{% choosable language python %}}

```python
import requests
headers = {
  'Accept': 'application/json',
  'Authorization': 'token pul-abc123'
}

> r = requests.get('https://api.pulumi.com/api/orgs/{org}/search/resources?properties=true?query=".instanceType"', headers = headers)

Reviewer 3: Need to update the URL encoding for this query and others in your code samples.

    public async Task GetAsync(string url)
    {
        //Start the request
        HttpResponseMessage response = await Client.GetAsync(url);

        //Validate result
        response.EnsureSuccessStatusCode();

    }

    /// Deserialize object from request response
    private async Task DeserializeObject(HttpResponseMessage response)
    {
        //Read body
        string responseBody = await response.Content.ReadAsStringAsync();

        //Deserialize Body to object
        var result = JsonConvert.DeserializeObject(responseBody);
    }
}

```

{{< /choosable >}}

{{% choosable language java %}}

```java
URL obj = new URL("https://api.pulumi.com/api/orgs/{org}/search/resources");
HttpURLConnection con = (HttpURLConnection) obj.openConnection();
con.setRequestMethod("GET");
int responseCode = con.getResponseCode();
BufferedReader in = new BufferedReader(
    new InputStreamReader(con.getInputStream()));
String inputLine;
StringBuffer response = new StringBuffer();
while ((inputLine = in.readLine()) != null) {
    response.append(inputLine);
}
in.close();
System.out.println(response.toString());

```

{{< /choosable >}}

{{< /chooser >}}

## Wrapping it all up

> We are excited to be unveiling Pulumi Property Search today and can’t wait to see what new capabilities it unlocks for your organization. As always, submit any feedback on the feature in the [Pulumi Cloud Requests](https://github.com/pulumi/pulumi-cloud-requests/issues/new?assignees=&labels=kind%2Fenhancement&projects=&template=feature-request.md) repository.

Reviewer 2: ```suggestion
We are excited to be extending Pulumi Resource Search with access to search over every property of your resources. We can’t wait to see what new capabilities it unlocks for your organization. As always, submit any feedback on the feature in the [Pulumi Cloud Requests](https://github.com/pulumi/pulumi-cloud-requests/issues/new?assignees=&labels=kind%2Fenhancement&projects=&template=feature-request.md) repository.
```

<!--more-->

Pulumi Resource Search, including the ability to search on property data, is applicable across all cloud resources, whether they reside in AWS, Google Cloud, Azure, or any of Pulumi's 100+ cloud providers. Support for querying on property data is available in the [Enterprise and Business Critical editions](/pricing) of Pulumi Cloud.

### See it in action

![Searching on resource properties in the Pulumi Cloud console](properties.gif)

## Feature Overview

> Cloud resources, regardless of the provider, come with various properties that define their behavior, performance, location, and more. You can now use Pulumi Resource Search to search resources based on these properties in addition to other resource metadata. Whether you're trying to find resources running in a specific region, instances with a particular tag, or databases running a certain engine version, searching over properties with Pulumi Cloud streamlines and simplifies your tasks.

Reviewer 4: ```suggestion
Cloud resources come with various properties that define their behavior, performance, location, and more. You can now use Pulumi Resource Search to search based on these properties in addition to other resource metadata across 100+ clouds. Whether you're trying to find resources running in a specific region, instances with a particular tag, or databases running a certain engine version, searching over properties with Pulumi Cloud streamlines and simplifies your tasks.
```

Using the new feature is simple and straightforward. Go to the Pulumi console, navigate to the search bar, and start typing the property you're interested in. For example, if you're searching for all AWS RDS instances running MySQL with a particular engine version, simply type `.engineVersion:5.7` and press enter.

The ability to search for resources based on their properties is a powerful tool for optimizing cloud resource management. It provides deeper insights into your infrastructure and aids in identifying resources that are misconfigured, out-of-date, or not adhering to your organization's best practices. Essentially, it's a significant step towards more proactive, informed, and efficient management of your cloud environment.

> Since launching Resource Search, a majority of Pulumi customers have used it, making it the most quickly adopted Pulumi feature. One of these customers is Alkria, a company that reinvents networking for the cloud era with global unified network infrastructure delivered as-a-service. 

Reviewer 4: I'd delete this sentance and change the previous sentence to be:

```
Customers like Alkira, a company that reinvents networking for the cloud era, have found it to be a significant step towards more proactive, informed, and efficient management of the cloud.
```

then right on to the quote

> “I’m making developers at Alkira significantly more productive while also making my job easier using Pulumi’s IaC platform and features like Pulumi Insights and Deployments,” said Santosh Dornal, Head of Software Test & DevOps, Alkira. “I can get developers using IaC immediately with Pulumi Deployments and its GitHub integration, while Pulumi Insights makes it really easy to find idle developer environments that need to be shut down, which reduces our cloud costs.”

## Example queries

> Before Property Search, using just Resource Search, you were able to find resources by their name, type, URN, parent, the team they are in, and so on... but today, any property on your resource will be indexed by Pulumi Cloud and searchable after each stack update. Let's walk through some examples of the types of queries you can now perform.

Reviewer 4: ```suggestion
Before Property Search, using just Resource Search, you were able to find resources by their name, type, URN, parent, the team they are in, and so on... but today, any input or output property on your resource will be indexed by Pulumi Cloud and searchable after each stack update. Let's walk through some examples of the types of queries you can now perform.
```

### AI Assist & properties

> Our [AI Assist](/product/private-previews) functionality allows you to use natural language to search for resources. AI Assist works by converting natural language to a query in our search syntax. This functionality becomes even more valuable with property search. Resources can have dozens of properties, and remembering exactly how to search for property values are can be challenging, especially when they change depending on the provider. AI Assist is going to step in so that you don’t have to remember everything.

Reviewer 4: ```suggestion
Pulumi [AI Assist](/product/private-previews) allows you to use natural language to search for resources. AI Assist works by converting natural language to a query in our search syntax. This becomes even more valuable with property search. Resources can have dozens of properties, and remembering exactly how to search for property values can be challenging, especially when they change depending on the provider. AI Assist is going to step in so that you don’t have to remember everything.
```

> Here is a list of example queries we have seen customers use:

Reviewer 4: ```suggestion
Here is a list of example natural language queries we have seen customers use and the AI Assist output:
```
